core/include/gnuradio-4.0/Port.hpp

```diff
@@ -955,7 +956,7 @@ template<typename T>
 concept TagPredicate = requires(const T &t, const Tag &tag, Tag::signed_index_type readPosition) {
     { t(tag, readPosition) } -> std::convertible_to<bool>;
 };
-inline constexpr TagPredicate auto defaultTagMatcher    = [](const Tag &tag, Tag::signed_index_type readPosition) noexcept { return tag.index >= readPosition || tag.index < 0; };
```

Comment:
Why the drop of `tag.index < 0`? This used to indicate that a tag should be applied immediately regardless of other index constraints.

Reply:
This predicate was and is used to get the next matching tag and the matching tag after that in `nSamplesUntilNextTag(port, offset=0)`, where offset is the default of 0 for getting the first tag and 1 for getting the Tag not at position 1. But since `offset` ends up in `readPosition` of the predicate tag index only, as soon a there is  a tag with index < 0, it will also be return when looking for the second tag... I did not follow up more detailed how this worked previously, re-adding it makes qa_Block spin indefinitely.

Tag propagation and setting _cachedTag is not influenced by this since there the `-1` condition is checked separately in `Port::getTags(untilOffset)`.

I'm not completely sure on whether this is the correct change, this is just the rationale for removing it. I tried to keep the changes outside of the workInternal (and its subfunctions) small, as i've been bitten sometimes by doing extensive changes thoughout the codebase at too many places at once.